apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -467,6 +471,25 @@ terminate(_Reason, _Session = #{}) ->
 
 %%--------------------------------------------------------------------
 
+make_reply_fun(ClientInfo, Subs, UpgradeQoS, InnerFun) ->
+    fun(Seqno, Message0 = #message{topic = Topic}) ->
+        emqx_utils:flattermap(
```

Comment:
I think this function can actually be implemented as fold that returns a tuple containing:
1) New inflight object
2) Number of QoS0 messages 
3) Number of QoS1 messages
4) Number of QoS2 messages

This will be helpful for improving the flow control later.